annotation/tests/test_tasks_services.py

```diff
@@ -295,3 +338,983 @@ def test_validate_user_actions_invalid_states(
+    mock_session.query.return_value = mock_query
+    mock_query.filter.return_value = mock_query
+    mock_query.first.return_value = mock_task
+
+    result = get_task_info(mock_session, task_id, tenant)
+
+    mock_session.query.assert_called_once_with(ManualAnnotationTask)
+    mock_query.first.assert_called_once()
+    assert result == mock_task
+
+
+@pytest.fixture
+def annotated_file_pages() -> List[int]:
+    return [1, 2, 3]
+
+
+def test_unblock_validation_tasks(
+    mock_session: Mock,
+    mock_task: ManualAnnotationTask,
+    annotated_file_pages: List[int],
+):
+    mock_unblocked_tasks = MagicMock()
+    mock_session.query.return_value.filter.return_value = mock_unblocked_tasks
+    mock_unblocked_tasks.all.return_value = [mock_task]
+
+    result = unblock_validation_tasks(
+        mock_session, mock_task, annotated_file_pages
+    )
+
+    mock_session.query.assert_called_once_with(ManualAnnotationTask)
+    mock_session.query.return_value.filter.assert_called_once()
+    mock_unblocked_tasks.update.assert_called_once_with(
+        {"status": TaskStatusEnumSchema.ready},
+        synchronize_session=False,
+    )
+    assert result == [mock_task]
+
+
+def test_get_task_stats_by_id(mock_session):
+    task_id = 1
+    mock_stats = MagicMock(spec=AnnotationStatistics)
+
+    mock_query = MagicMock()
+    mock_session.query.return_value = mock_query
+    mock_query.filter.return_value.first.return_value = mock_stats
+
+    result = get_task_stats_by_id(mock_session, task_id)
+
+    mock_session.query.assert_called_once_with(AnnotationStatistics)
+    mock_query.filter.assert_called_once()
+    mock_query.filter.return_value.first.assert_called_once()
+
+    assert result == mock_stats
+
+
+def test_add_task_stats_record_existing_stats(mock_session):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
+    mock_stats_input.event_type = "open"
+    mock_stats_db = MagicMock(spec=AnnotationStatistics)
+    with patch(
+        "annotation.tasks.services.get_task_stats_by_id"
+    ) as mock_get_task_stats_by_id:
+        mock_get_task_stats_by_id.return_value = mock_stats_db
+
+        result = add_task_stats_record(mock_session, task_id, mock_stats_input)
+
+        mock_get_task_stats_by_id.assert_called_once_with(
+            mock_session, task_id
+        )
+        mock_stats_db.field1 = "value1"
+        mock_stats_db.field2 = "value2"
+        mock_stats_db.updated = datetime.utcnow()
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+
+        assert result == mock_stats_db
+
+
+def test_add_task_stats_record_(mock_session):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
+    mock_stats_input.event_type = "closed"
+    with patch(
+        "annotation.tasks.services.get_task_stats_by_id"
+    ) as mock_get_task_stats_by_id:
+        mock_get_task_stats_by_id.return_value = None
```

Comment:
```suggestion
        "annotation.tasks.services.get_task_stats_by_id",
        return_value = None
    ) as mock_get_task_stats_by_id:
```
We can define the return_value inside of patch()